ArduPlane/mode_loiter.cpp

```diff
@@ -25,3 +26,46 @@ void ModeLoiter::update()
     plane.calc_throttle();
 }
 
+bool ModeLoiter::headingLinedUp(const Location loiterCenterLoc, const Location targetLoc)
+{
+    const uint16_t loiterRadius = abs(plane.aparm.loiter_radius);
+    if (loiterCenterLoc.get_distance(targetLoc) < loiterRadius + loiterRadius*0.05) {
+        /* Whenever next waypoint is within the loiter radius plus 5%,
+           maintaining loiter would prevent us from ever pointing toward the next waypoint.
+           Hence break out of loiter immediately
+         */
+        return true;
+    }
+
+    // Bearing in degrees
+    const int32_t bearing_cd = plane.current_loc.get_bearing_to(targetLoc);
+
+    // get current heading.
+    const int32_t heading_cd = plane.gps.ground_course_cd();
+
+    const int32_t heading_err_cd = wrap_180_cd(bearing_cd - heading_cd);
+
+    /*
+      Check to see if the the plane is heading toward the land
+      waypoint. We use 20 degrees (+/-10 deg) of margin so that
+      we can handle 200 degrees/second of yaw.
+
+      After every full circle, extend acceptance criteria to ensure
+      aircraft will not loop forever in case high winds are forcing
+      it beyond 200 deg/sec when passing the desired exit course
+    */
+
+    // Use integer division to get discrete steps
+    const int32_t expanded_acceptance = 1000 * (plane.loiter.sum_cd / 36000);
```

Comment:
plane.loiter.sum_cd can be negative

Reply:
some random guy from the internet fixed it!
https://github.com/magicrub/ardupilot/commit/1fcfc9a35cd1e98fc9694b241d25a4f4772f19b6